File: 2022/2022W09/PreppindataW09.py
import pandas as pd
import re

classification = {'N':'New', 'C':'Consistent', 'S':'Sleeping', 'R':'Returning'}

xlsx = pd.ExcelFile(r"C:\Data\PreppinData\Sample - Superstore.xls")

sales = pd.read_excel(xlsx,'Orders')
sales['Year'] = sales['Order Date'].apply(lambda x: x.year)

customer = sales.groupby(['Customer ID', 'Customer Name', 'Year'], as_index=False).agg({'Order ID':'count'})
customer['First Purchase'] = customer['Year'].groupby(customer['Customer ID']).transform('min')
customer = customer.pivot_table(index=['Customer ID', 'Customer Name', 'First Purchase'],
                                columns='Year', values='Order ID', aggfunc=max).reset_index()
customer = customer.melt(id_vars=['Customer ID', 'Customer Name', 'First Purchase'], value_name='Order?', var_name='Year')
customer['Order?'] = customer['Order?'].apply(lambda x: 0 if pd.isnull(x) else 1) 
customer['Position'] = customer['Year'] - customer['Year'].min()
customer['Power'] = customer['Year'].max() - customer['Year']
customer['Temp'] = customer['Order?']*(10**customer['Power']) 
customer['OrderString'] = customer['Temp'].groupby(customer['Customer ID']).transform('sum')
customer['OrderString'] = customer['OrderString'].apply(lambda x: re.sub('0', 'S', re.sub('(?<=0)(1)', 'R', re.sub('(?<!0)(1)', 'C', re.sub('(^1)', 'N', str(x))))).rjust(4, '-'))
customer['Customer Classification'] = customer.apply(lambda x: classification.get(x['OrderString'][x['Position']],'-'), axis=1)
customer = customer[customer['Customer Classification']!='-'][['Customer ID', 'Customer Name', 'First Purchase', 'Year', 'Order?', 'Customer Classification']]

cohort = customer.groupby(['First Purchase', 'Year'], as_index=False).agg({'Order?':'sum'})
cohort = cohort.sort_values(['First Purchase', 'Year'])
cohort['Previous'] = cohort.groupby('First Purchase')['Order?'].transform(lambda x: x.shift(1))
cohort['YoY Difference'] = cohort.apply(lambda x: (x['Order?'] - x['Previous']) if x['Year'] > x['First Purchase'] else None, axis=1 )

customer = pd.merge(customer, cohort[['First Purchase', 'Year', 'YoY Difference']], on=['First Purchase', 'Year'])
final = pd.merge(customer, sales, on=['Customer ID', 'Customer Name', 'Year'], how='left')

File: 2022/2022W09/preppindataW09.R
library(readxl)
library(dplyr)
library(tidyr)
library(lubridate)
library(stringr)

xlsx <- "C:/Data/PreppinData/Sample - Superstore.xls"

status_transfer <- function(order_str){
  ret <- str_replace_all(order_str, '(^1)', 'N') %>%
    str_replace_all(., '(?<!0)(1)', 'C') %>%
    str_replace_all(., '(?<=0)(1)', 'R') %>%
    str_replace_all(., '0', 'S') %>%
    str_pad(., 4, 'left', '-')
  return(ret)}

customer_class <- function(class_code){
  ret <- case_when(
    class_code == 'N' ~ 'New',
    class_code == 'C' ~ 'Consistent',
    class_code == 'S' ~ 'Sleeping',
    class_code == 'R' ~ 'Returning',
    TRUE ~ '-' )
  return(ret)}

sales <- read_excel(xlsx, 'Orders') %>% mutate('Year' = year(`Order Date`))

customer <- sales %>%
  select(c('Customer ID', 'Customer Name', 'Year')) %>%
  distinct() %>%
  group_by(`Customer ID`) %>%
  mutate('Order?' = 1,
         'First Purchase' = min(`Year`)) %>%
  ungroup() %>%
  pivot_wider(names_from = 'Year', values_from = 'Order?', values_fill = 0) %>%
  pivot_longer(cols=matches('^\\d'), names_to = 'Year', values_to = 'Order?', names_transform = list('Year'=as.integer)) %>%
  group_by(`Customer ID`) %>%
  mutate('Position' = `Year` - min(`Year`) + 1,
         'Power' = max(`Year`) - `Year`, 
         'OrderString' = status_transfer(as.character(sum(`Order?`*10^`Power`))),
         'Customer Classification' = customer_class(substr(`OrderString`, `Position`, `Position`))) %>%
  filter(`Customer Classification` != '-') %>%
  select(c('Customer ID', 'Customer Name', 'First Purchase', 'Year', 'Order?', 'Customer Classification'))
  
cohort <- customer %>%
  group_by(`First Purchase`, `Year`) %>%
  summarise('Order?' = sum(`Order?`), .groups='drop') %>%
  arrange(`First Purchase`, `Year`) %>%
  group_by(`First Purchase`) %>%
  mutate('YoY Difference' = `Order?` - lag(`Order?`)) %>%
  select(c('First Purchase', 'Year', 'YoY Difference'))
  
final <- merge(customer, cohort, by=c('First Purchase', 'Year')) %>%
  merge(., sales, by=c('Customer ID', 'Customer Name', 'Year'), all.x=TRUE)

View(final)